pennylane/_qubit_device.py

```diff
@@ -1886,3 +1886,138 @@ def adjoint_jacobian(self, tape, starting_state=None, use_device_state=False):
+
+        # Initialization of state
+        if starting_state is not None:
+            ket = self._reshape(starting_state, [2] * self.num_wires)
+        else:
+            if not use_device_state:
+                self.reset()
+                self.execute(tape)
+            ket = self._pre_rotated_state
+
+        n_obs = len(tape.observables)
+        bras = np.empty([n_obs] + [2] * self.num_wires, dtype=np.complex128)
+        for kk in range(n_obs):
+            bras[kk, ...] = self._apply_operation(ket, tape.observables[kk])
+
+        expanded_ops = []
+        for op in reversed(tape.operations):
+            if op.num_params > 1:
+                if isinstance(op, qml.Rot) and not op.inverse:
+                    ops = op.decomposition()
+                    expanded_ops.extend(reversed(ops))
+                else:
+                    raise qml.QuantumFunctionError(
+                        f"The {op.name} operation is not supported using "
+                        'the "adjoint" differentiation method'
+                    )
+            else:
+                if op.name not in ("QubitStateVector", "BasisState", "Snapshot"):
+                    expanded_ops.append(op)
+
+        trainable_params = []
+        for k in tape.trainable_params:
+            # pylint: disable=protected-access
+            if hasattr(tape._par_info[k]["op"], "return_type"):
+                warnings.warn(
+                    "Differentiating with respect to the input parameters of "
+                    f"{tape._par_info[k]['op'].name} is not supported with the "
+                    "adjoint differentiation method. Gradients are computed "
+                    "only with regards to the trainable parameters of the circuit.\n\n Mark "
+                    "the parameters of the measured observables as non-trainable "
+                    "to silence this warning.",
+                    UserWarning,
+                )
+            else:
+                trainable_params.append(k)
+
+        jac = np.zeros((len(tape.observables), len(trainable_params)))
+
+        param_number = len(tape.get_parameters(trainable_only=False, operations_only=True)) - 1
+        trainable_param_number = len(trainable_params) - 1
+        for op in expanded_ops:
+
+            adj_op = qml.adjoint(op)
+            ket = self._apply_operation(ket, adj_op)
+
+            if op.grad_method is not None:
+                if param_number in trainable_params:
+                    d_op_matrix = operation_derivative(op)
+                    ket_temp = self._apply_unitary(ket, d_op_matrix, op.wires)
+
+                    jac[:, trainable_param_number] = 2 * dot_product_real(bras, ket_temp)
+
+                    trainable_param_number -= 1
+                param_number -= 1
+
+            for kk in range(n_obs):
+                bras[kk, ...] = self._apply_operation(bras[kk, ...], adj_op)
+
+        if len(tape.observables) == 1:
+            return jac[0]
+
+        return tuple(jac)
```

Comment:
Are these lines the only diff from `adjoint_jacobian`? If so, `adjoint_jacobian_new` could call `adjoint_jacobian` and post-process its result to ensure an ease of maintainability. This would be a fortunate case until no other part of the logic requires change.

Reply:
I'm curious if it should be the other way around? `adjoint_jacobian` checks if the new return type is enabled, and instead calls `adjoint_jacobian_new`/does post-processing?

BTW we should probably name it something more descriptive, like `adjoint_jacobian_tuple` 🙂

Reply:
> I'm curious if it should be the other way around? adjoint_jacobian checks if the new return type is enabled, and instead calls adjoint_jacobian_new/does post-processing?

I think that works great too. :+1: :slightly_smiling_face: 

> BTW we should probably name it something more descriptive, like adjoint_jacobian_tuple slightly_smiling_face

Very fair point! We have some `_new` functions in the code base already, so likely best to come to a decision on naming for consistency. We'll be chatting about this in more depth via calls, saving this point for discussion and will return here.

Reply:
Ah got it! If it is doable, I think having `_tuple` might make more sense? But only for user/external-developer facing ones. Internal implementations I leave up to you :)

Reply:
It does :) It may just be that not all such functions return `_tuple` (for jacobians they should, however).

We've converged on a convention on code organization. Having said that, from my point of view, it's okay not to strictly follow it here because the changes required seem straightforward enough as they are now.